### DISCUSSION TITLE
Single Object Tracking using YOLOv8 trained model

### DISCUSSION THREAD

**Discussion by 4833R11Y45** [ultralytics/ultralytics#8267] [Q&A]
Hi,
It has been almost 8 months since I was introduced to Computer Vision, and eventually started out with YOLOv8, and things have never been the same before!
Thus, I have chosen YOLOv8 as my major object detection model for my FYP as well as other projects.
In my FYP, I am trying to implement an object tracker that lets the user to track specific vehicle (my trained YOLOv8 model's target object) in the video frame provided by my custom-built drone. Like for example, say there are 10 cars in the frame that are being detected and the user wants to track only one specific vehicle. For that purpose, I have decided to use Tracking ID as the selection criteria. So, what I want is when is that once a vehicle is selected, then only that vehicle is highlighted, and we could receive that vehicle's midpoint coordinates. So, it will be really helpful if someone can guide me with this situation, I am really stuck in this situation for the past 3 weeks as no person is university is holding expertise in Computer Vision and there's no relevant guide about it on the internet.
Thank you in advance.

**Comment by pderrenger** [comment]
Hi there! 👋

It's great to hear about your enthusiasm for Computer Vision and your choice to use YOLOv8 for your FYP and other projects! For single object tracking using a YOLOv8 trained model, you can indeed use the tracking ID to follow a specific vehicle. Here's a simple approach you could take:

1. Use the `track` mode of YOLOv8 to get tracking IDs for all detected vehicles.
2. Implement a selection mechanism to choose the vehicle of interest based on its tracking ID.
3. Once selected, filter the tracking results to only update and highlight the chosen vehicle.

Here's a snippet of how you might implement this in Python:

```python
from ultralytics import YOLO

# Load your custom trained model
model = YOLO('path/to/your/custom_model.pt')

# Run tracking on your video
results = model.track(source='path/to/your/video.mp4', stream=True)

# Assuming you have a function to select the vehicle ID of interest
vehicle_id_of_interest = select_vehicle_id()

for result in results:
    # Filter results for the vehicle of interest
    if result.boxes.id == vehicle_id_of_interest:
        # Get midpoint coordinates (x_center, y_center)
        midpoint = result.boxes.xywh[0][0:2]
        print(f"Tracking vehicle ID {vehicle_id_of_interest} at {midpoint}")

        # Here you can add code to highlight the vehicle in the frame
        # ...
```

Remember to replace `select_vehicle_id()` with your actual selection mechanism, and `path/to/your/custom_model.pt` and `path/to/your/video.mp4` with the correct paths to your model and video file.

I hope this helps you move forward with your FYP! If you have any more questions or need further assistance, feel free to ask. Good luck with your project! 🚀

**Reply by 4833R11Y45** [reply]
Thank you so much for this, I used this as the base code and built up the logic as needed. My supervisor was happy with that. I gave the demo to the FYP panel, and they showed a sort of concern about the deployment of the model as well as of the tracker on the on-board processor of the UAV (Orange Pi 5, 16 GB RAM, No GPU, Rockchip RK3588S 64-bit processor with 8 cores).

I analyzed my trained model on my laptop (Core i5 13 gen, 24 GB Ram, NVIDIA RTX 3050 6 GB) that it utilizes 3 GB RAM and 0.5 to 1 GB VRAM (upto 30% utilization indication) GPU during inference with inference speed = 20ms in Pytorch framework. And on the other hand side, the same model in NCNN format utilizes 0.7 GB RAM and 0.1 GB of GPU with the inference speed = 172ms. So, you can see it is way slow than the Pytorch framework and won't be helpful in real-time scenario.

Is there any way to increase the inference speed of NCNN model in such a way that it can be used for real-time detection and tracking using Orange Pi 5, or any other way for the real time detection and tracking on Orange Pi 5, I will be really thankful if Ultralytics tea can guide me in this situation. Thank you in advance.

**Reply by pderrenger** [reply]
Hi there! 🌟

It's fantastic to hear that your base code worked well and impressed your supervisor and the FYP panel! Deploying models on devices like the Orange Pi 5 without a GPU does present unique challenges, especially for real-time applications.

Given the constraints of the Orange Pi 5 and the performance you've observed with the NCNN format, optimizing for speed becomes crucial. Here are a couple of suggestions:

1. **Model Simplification**: Consider simplifying your model if possible. This might involve reducing the number of layers or parameters, which can significantly decrease inference time at the cost of some accuracy.

2. **Quantization**: Applying quantization can reduce model size and speed up inference by using lower-precision (e.g., INT8 vs. float32) calculations. NCNN supports quantization, which could help in your case.

3. **NCNN Optimization**: Make sure you're leveraging all NCNN optimizations, such as using Vulkan for computation if supported by your hardware, and optimizing the model specifically for the ARM architecture of the Orange Pi.

4. **Batch Processing**: If real-time constraints allow, processing multiple frames at once (batch processing) can improve overall throughput.

5. **Efficient Tracking**: For tracking, consider lightweight algorithms that require minimal computation. Sometimes, simpler algorithms like centroid tracking can be surprisingly effective and fast.

Here's a quick example of how you might apply quantization with NCNN (assuming you have the tools set up):

```bash
# Convert the model to NCNN format with quantization
./ncnn/tools/quantize/ncnn2table your_model.param your_model.bin
./ncnn/tools/quantize/quantize your_model.param your_model.bin your_model.table
```

Remember, achieving real-time performance on hardware like the Orange Pi 5 is challenging but not impossible. It often requires a balance between speed, accuracy, and model complexity.

I hope these suggestions help you optimize your model for the Orange Pi 5! If you have further questions or need more detailed guidance, feel free to reach out. Good luck with your project! 🚀
